### PR TITLE
EC-802 Include SHA/Image Digest in report output

### PR DESCRIPTION
**Pull request by robnester-rh**
This commit adds functionality to display the SHA/Image digest in the report output. Speicifically it ensures that for git and oci sources, the identifying SHA/Image Digests are determined and provided in the report output, allowing for auditing and reproducibility.

### CODE REVIEW

**Comment by robnester-rh** [comment]
There are four files at the core of this PR:
```
cmd/validate/image.go
internal/cache/cache.go
internal/policy/source/source.go
internal/validate/helpers.go
```
The remainder of the churn in this PR is due to updating tests and snap files as we now return the pinned SHA / image digest in the output.

File: internal/policy/source/source.go [hunk 1]
```diff
@@ -170,6 +189,45 @@ func (p *PolicyUrl) Subdir() string {
 	return string(p.Kind)
 }
 
+func (p PolicyUrl) Type() PolicyType {
+	return p.Kind
+}
+
+func getPinnedUrl(u string, m metadata.Metadata) (string, error) {
```
Comment (lcarva) [inline internal/policy/source/source.go:196]: I hope that eventually this moves to go-gather itself. Could you add a TODO comment if you agree?

File: internal/validate/helpers.go [hunk 2]
```diff
@@ -79,3 +90,85 @@ func ReadFile(ctx context.Context, fileName string) (string, error) {
 	log.Debugf("Loaded %s", fileName)
 	return string(fileBytes), nil
 }
+
+// PreProcessPolicy fetches policy sources and returns a policy object with
+// pinned SHA/image digest URL where applicable, along with a policy cache object.
+func PreProcessPolicy(ctx context.Context, policyOptions policy.Options) (policy.Policy, *cache.PolicyCache, error) {
```
Comment (lcarva) [inline internal/validate/helpers.go:96]: This should probably go under `internal/policy`?

File: internal/validate/helpers.go [hunk 3]
```diff
@@ -79,3 +90,85 @@ func ReadFile(ctx context.Context, fileName string) (string, error) {
...
+	var policyCache *cache.PolicyCache
+	pinnedPolicyUrls := map[string][]string{}
+	policyCache, ok := PolicyCacheFromContext(ctx)
```
Comment (lcarva) [inline internal/validate/helpers.go:99]: This is a similar pattern to how we create _clients_ that make external connections. Consider introducing a helper function, `NewPolicyCache` that handles the logic of using an existing cache from the context, or creating a new one:  https://github.com/enterprise-contract/ec-cli/blob/0c2bc064788fbeacec82b7492e7330832a9b41d1/internal/kubernetes/client.go#L59-L74

File: internal/validate/helpers.go [hunk 4]
```diff
@@ -79,3 +90,85 @@ func ReadFile(ctx context.Context, fileName string) (string, error) {
...
+				log.Debugf("Unable to download source from %s!", policySource.PolicyUrl())
+				return nil, nil, err
+			}
+			log.Debugf("Downloaded policy source from %s to %s\n", policySource.PolicyUrl(), destDir)
+
+			url := policySource.PolicyUrl()
+
+			if _, found := policyCache.Get(policySource.PolicyUrl()); !found {
+				log.Debugf("Cache miss for: %s, adding to cache", url)
+				policyCache.Set(url, destDir, nil)
```
Comment (lcarva) [inline internal/validate/helpers.go:144]: May want to consider adding the pinned version of the `url` to the cache as well.

**Review by lcarva — COMMENTED** [review]
I think most of my comments are nitpicks that can be addressed later on. I'm ok with merging so long as there's an immediate followup PR to address some of them.

**Review by zregvart — APPROVED** [review]
LGTM, look at the comments Luiz made and see if we want to do a followup or a touch-up here

**Comment by robnester-rh** [comment]
Added touch-ups addressing all comments. Thank you for the reviews.

File: cmd/validate/image_integration_test.go [hunk 5]
```diff
@@ -44,15 +45,19 @@ import (
...
 	ctx = oci.WithClient(ctx, &client)
-
-	noEvaluators := 100
+	mdl := MockDownloader{}
+	downloaderCall := mdl.On("Download", mock.Anything, mock.Anything, false).Return(&ociMetadata.OCIMetadata{Digest: "sha256:da54bca5477bf4e3449bc37de1822888fa0fbb8d89c640218cb31b987374d357"}, nil).Times(noEvaluators)
+	ctx = context.WithValue(ctx, source.DownloaderFuncKey, &mdl)
 
 	evaluators := make([]*mockEvaluator, 0, noEvaluators)
-	expectations := make([]*mock.Call, 0, noEvaluators)
+	expectations := make([]*mock.Call, 0, noEvaluators+1)
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply cmd/validate/image_integration_test.go:59]: Yes.

File: cmd/validate/image_integration_test.go [hunk 6]
```diff
@@ -67,7 +72,7 @@ func TestEvaluatorLifecycle(t *testing.T) {
 	}
 
 	newConftestEvaluator = func(_ context.Context, s []source.PolicySource, _ evaluator.ConfigProvider, _ v1alpha1.Source) (evaluator.Evaluator, error) {
-		idx, err := strconv.Atoi(s[0].PolicyUrl())
+		idx, err := strconv.Atoi(strings.Split(strings.Split(s[0].PolicyUrl(), "@")[0], "://")[1])
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply cmd/validate/image_integration_test.go:76]: Done

File: internal/cache/cache.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	return entry.value, ok
+}
+
+// Set manually sets the value for a given key in the cache.
+// It overwrites any existing value and error.
+func (c *PolicyCache) Set(key string, value string, err error) {
+	entry := &cacheEntry{
+		value: value,
+		err:   err,
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/cache/cache.go]: The idea is that we perform a sync.OnceValue to keep from having to download multiple types, if we've hit an error once, we already have a failure so we don't try to download again when we return the failure from a `.Get()` call.

File: internal/evaluator/conftest_evaluator.go [hunk 8]
```diff
@@ -294,19 +294,15 @@ func NewConftestEvaluatorWithNamespace(ctx context.Context, policySources []sour
...
 	if err != nil {
 		log.Debug("Failed to create work dir!")
 		return nil, err
 	}
 	c.workDir = dir
-
 	c.policyDir = filepath.Join(c.workDir, "policy")
 	c.dataDir = filepath.Join(c.workDir, "data")
 
-	log.Debugf("Created work dir %s", dir)
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/evaluator/conftest_evaluator.go:308]: Correct, I've put this back.

File: internal/policy/source/source.go [hunk 9]
```diff
@@ -144,8 +147,14 @@ func getPolicyThroughCache(ctx context.Context, s PolicySource, workDir string,
 		if _, ok := c.metadata.(*ociMetadata.OCIMetadata); ok {
 			log.Debugf("Image digest for source(%s): %s\n", s.PolicyUrl(), c.metadata.(*ociMetadata.OCIMetadata).Digest)
 		}
+		if _, ok := c.metadata.(*fileMetadata.FileMetadata); ok {
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/policy/source/source.go]: Extracted this to a small function that uses type switch statement.

File: internal/validate/helpers.go (hunk 2 above)
Comment (robnester-rh) [reply internal/validate/helpers.go]: Moved.

File: internal/validate/helpers.go (hunk 3 above)
Comment (robnester-rh) [reply internal/validate/helpers.go]: Done.

File: internal/validate/helpers.go (hunk 4 above)
Comment (robnester-rh) [reply internal/validate/helpers.go]: I'll revisit this once I move the `pinnedUrl function`.

File: internal/policy/source/source.go (hunk 1 above)
Comment (robnester-rh) [reply internal/policy/source/source.go:178]: Added.

File: internal/cache/cache.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package cache
```
[the remark this answers is not in the document]
Comment (robnester-rh) [reply internal/cache/cache.go]: Moved.

**Review by zregvart — APPROVED** [review]

**Pull request merged**